File: medium/0659-encode-and-decode-strings.js
class Solution {
  /**
   * @param {string[]} strs
   * @returns {string}
   */
  // This is more efficient than having output = '' and doing
  // output += str.length + '#' + str each iteration of strs.
  encode(strs) {
    let output = [];

    strs.forEach((str) => {
      output.push(str.length);
      output.push('#');
      output.push(str);
    });

    return output.join('');
  }

  /**
   * @param {string} str
   * @returns {string[]}
   */
  decode(str) {
    const output = [];
    let i = 0;

    while (i < str.length) {
      let j = i;

      while (str[j] != '#') {
        j++;
      }

      // j will be pointing at #
      const wordLen = parseInt(str.slice(i, j));
      output.push(str.slice(j + 1, j + 1 + wordLen));
      i = j + 1 + wordLen;
    }

    return output;
  }
}

// Time complexity: O(n) for encode and decode
// Space complexity: O(1) for encode and decode
// where n is the sum of the lengths of all strings in strs

File: medium/0659-encode-and-decode-strings.py
from typing import List


class Solution:
    # If m is the sum of lens of all the strings
    # and n is the total number of strings
    # Time complexity: O(m)
    # Space complexity: O(m + n)
    def encode(self, strs: List[str]) -> str:
        res = []
        for s in strs:
            # Encode by prepending len of str and separator to str
            res.append(f"{len(s)}#{s}")
        return "".join(res)

    # If m is the sum of lens of all the strings
    # and n is the total number of strings
    # Time complexity: O(m)
    # Space complexity: O(m + n)
    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        while i < len(s):
            # Read len of curr str until separator
            j = i + 1
            while s[j] != "#":
                j += 1
            strLen = int(s[i:j])

            # Slicing [i:j] will give us curr str
            i = j + 1
            j = i + strLen
            res.append(s[i:j])
            i = j
        return res
